<a href="https://colab.research.google.com/github/AI-Krittaya/BSC_DPDM2025/blob/main/Midterm_663020277_7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Import Data

In [14]:
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [28]:
historical  = pd.read_csv('/content/drive/MyDrive/BSC DPDM midterm 2025/Daily rain - (Statistics 2012-2024)/hii_daily_rain.csv')

In [16]:
historical

,station_code,measure_datetime,data,station_name,latitude,longitude,quality_flag,basin,sub_basin,tambon,amphoe,province,data_type
0,ABRT,2012-01-01 07:00:00+07,0.0,วัดเวฬุวัน,16.054785,103.664290,U,ชี,ลำน้ำชีส่วนที่ 4/2,เหนือเมือง,เมืองร้อยเอ็ด,ร้อยเอ็ด,rainfall_daily
1,ABRT,2012-01-03 07:00:00+07,0.0,วัดเวฬุวัน,16.054785,103.664290,U,ชี,ลำน้ำชีส่วนที่ 4/2,เหนือเมือง,เมืองร้อยเอ็ด,ร้อยเอ็ด,rainfall_daily
2,ABRT,2012-01-06 07:00:00+07,0.0,วัดเวฬุวัน,16.054785,103.664290,U,ชี,ลำน้ำชีส่วนที่ 4/2,เหนือเมือง,เมืองร้อยเอ็ด,ร้อยเอ็ด,rainfall_daily
3,ABRT,2012-01-07 07:00:00+07,0.0,วัดเวฬุวัน,16.054785,103.664290,U,ชี,ลำน้ำชีส่วนที่ 4/2,เหนือเมือง,เมืองร้อยเอ็ด,ร้อยเอ็ด,rainfall_daily
4,ABRT,2012-01-08 07:00:00+07,0.0,วัดเวฬุวัน,16.054785,103.664290,U,ชี,ลำน้ำชีส่วนที่ 4/2,เหนือเมือง,เมืองร้อยเอ็ด,ร้อยเอ็ด,rainfall_daily
...,...,...,...,...,...,...,...,...,...,...,...,...,...
3768458,YOM016,2024-11-25 00:00:00+07,0.0,ห้วยแม่พวก,17.968712,100.063545,N,NaN,NaN,เด่นชัย,เด่นชัย,แพร่,rainfall_daily
3768459,YOM016,2024-11-26 00:00:00+07,0.0,ห้วยแม่พวก,17.968712,100.063545,N,NaN,NaN,เด่นชัย,เด่นชัย,แพร่,rainfall_daily
3768460,YOM016,2024-11-27 00:00:00+07,0.0,ห้วยแม่พวก,17.968712,100.063545,N,NaN,NaN,เด่นชัย,เด่นชัย,แพร่,rainfall_daily
3768461,YOM016,2024-11-28 00:00:00+07,0.0,ห้วยแม่พวก,17.968712,100.063545,N,NaN,NaN,เด่นชัย,เด่นชัย,แพร่,rainfall_daily


ความหมายของคอลัมน์ในตาราง historical

* station_code: รหัสสถานี หรือชื่อสถานีที่ใช้ระบุแต่ละสถานี
* measure_datetime: วันที่และเวลาที่มีการตรวจวัดข้อมูล
* data: ค่าข้อมูลที่ตรวจวัดได้ ในกรณีนี้อาจหมายถึงปริมาณน้ำฝน
* station_name: ชื่อของสถานีตรวจวัด
* latitude: ละติจูด (Latitude) แสดงตำแหน่งทางภูมิศาสตร์ในแนวนอน (เหนือ-ใต้)
* longitude: ลองจิจูด (Longitude) แสดงตำแหน่งทางภูมิศาสตร์ในแนวตั้ง (ตะวันออก-ตะวันตก)
* quality_flag: ธงระบุคุณภาพของข้อมูล (เช่น U = Unchecked, N = Normal)
* basin: ลุ่มน้ำที่สถานีตั้งอยู่
* sub_basin: ลุ่มน้ำย่อยที่สถานีตั้งอยู่
* tambon: ตำบลที่สถานีตั้งอยู่
* amphoe: อำเภอที่สถานีตั้งอยู่
* province: จังหวัดที่สถานีตั้งอยู่
* data_type: ชนิดของข้อมูลที่ตรวจวัด (เช่น rainfall_daily)

In [29]:
station = pd.read_csv('/content/drive/MyDrive/BSC DPDM midterm 2025/Input - (Operation Daily rain from API_ 2025)/Stations in the model_HII.csv')

In [23]:
station

,Node,Lat,Long
0,ACRU,15.788054,104.642235
1,BBHN,9.536913,98.579490
2,BBUA,17.372976,103.984436
3,BCNG,17.386840,103.290870
4,BDAR,14.587073,102.495190
...,...,...,...
380,YOM003,18.265924,100.177060
381,YOM008,16.708166,100.171780
382,YOM009,16.310750,100.271750
383,YOM010,18.133717,100.124596


ความหมายของคอลัมน์ Node Lat Long ในตาราง `station`

* Node: รหัสสถานี หรือชื่อสถานีที่ใช้ระบุแต่ละสถานี
* Lat: ละติจูด (Latitude) แสดงตำแหน่งทางภูมิศาสตร์ในแนวนอน (เหนือ-ใต้)
* Long: ลองจิจูด (Longitude) แสดงตำแหน่งทางภูมิศาสตร์ในแนวตั้ง (ตะวันออก-ตะวันตก)

In [30]:
import glob
import pandas as pd

allfiles = glob.glob('/content/drive/MyDrive/BSC DPDM midterm 2025/Input - (Operation Daily rain from API_ 2025)/**/*.csv', recursive=True)

# Filter out the stations file, as it does not contain 'measure_datetime'
rainfall_files = [f for f in allfiles if 'Stations in the model_HII.csv' not in f]

daily_2025 = pd.concat(
    [pd.read_csv(f, parse_dates=['measure_datetime']) for f in rainfall_files],
    ignore_index=True
)

daily_2025.head()

,rainfall_datetime,rainfall_datetime_calc,rainfall_value,tele_station_id,tele_station_oldcode,measure_datetime,station_code,value
0,2025-08-03 00:00:00+07:00,2025-08-03T07:00:00+07:00,0.0,575571.0,BKC007,2025-08-02,NaN,NaN
1,2025-08-03 00:00:00+07:00,2025-08-03T07:00:00+07:00,0.0,1512928.0,NaN,2025-08-02,NaN,NaN
2,2025-08-03 00:00:00+07:00,2025-08-03T07:00:00+07:00,0.0,1512927.0,NaN,2025-08-02,NaN,NaN
3,2025-08-03 00:00:00+07:00,2025-08-03T07:00:00+07:00,9.0,1395256.0,NaN,2025-08-02,NaN,NaN
4,2025-08-03 00:00:00+07:00,2025-08-03T07:00:00+07:00,0.0,1395255.0,NaN,2025-08-02,NaN,NaN


ความหมายของคอลัมน์ในตาราง `daily_2025`

*   **rainfall_datetime**: วันที่และเวลาที่มีการตรวจวัดปริมาณน้ำฝน (อาจรวมถึง Timezone offset)
*   **rainfall_datetime_calc**: วันที่และเวลาที่มีการคำนวณหรือปรับปรุงข้อมูลปริมาณน้ำฝน (อาจรวมถึง Timezone offset)
*   **rainfall_value**: ค่าปริมาณน้ำฝนที่ตรวจวัดได้ในหน่วยมิลลิเมตร (mm)
*   **tele_station_id**: รหัสประจำตัวของสถานีโทรมาตร (telemetry station)
*   **tele_station_oldcode**: รหัสเก่าของสถานีโทรมาตร
*   **measure_datetime**: วันที่และเวลาที่มีการตรวจวัดข้อมูลทั่วไป
*   **station_code**: รหัสสถานีที่ใช้ระบุแต่ละสถานี (อาจเป็นรหัสเดียวกับ `tele_station_oldcode` หรือ `tele_station_id`)
*   **value**: ค่าข้อมูลที่ตรวจวัดได้ (ในบริบทนี้อาจเป็นปริมาณน้ำฝนเช่นเดียวกับ `rainfall_value` หรือเป็นค่าอื่น ๆ ขึ้นอยู่กับข้อมูล)

In [31]:
Output = pd.read_csv('/content/drive/MyDrive/BSC DPDM midterm 2025/Output - (Clean monthly rain - 2025)/Y_monthly_rain_HII_operation.csv')

In [9]:
Output

,Node,rain,month,year
0,ACRU,83.0,7,2025
1,ACRU,164.8,8,2025
2,ACRU,248.6,9,2025
3,ACRU,48.4,10,2025
4,ACRU,52.2,11,2025
5,ACRU,0.2,12,2025
6,BBHN,391.6,7,2025
7,BBHN,641.2,8,2025
8,BBHN,517.6,9,2025
9,BBHN,198.8,10,2025


**คำอธิบายคอลัมน์ในตาราง `Output`**

*   **Node**: รหัสสถานี หรือชื่อสถานีที่ใช้ระบุแต่ละสถานี
*   **rain**: ปริมาณน้ำฝนรวมรายเดือน (มิลลิเมตร) ที่สถานีนั้นๆ ได้รับ
*   **month**: เดือนของข้อมูลปริมาณน้ำฝน (เป็นตัวเลข 1-12)
*   **year**: ปีของข้อมูลปริมาณน้ำฝน

#กรองข้อมูลสถานี

In [32]:
before = daily_2025['tele_station_oldcode'].nunique()

daily_2025 = daily_2025[
    daily_2025['tele_station_oldcode'].astype(str).isin(station['Node'].astype(str))
]

after = daily_2025['tele_station_oldcode'].nunique()

print(f"จำนวนสถานีก่อนกรอง: {before}")
print(f"จำนวนสถานีหลังกรอง: {after}")

print("\nตัวอย่างข้อมูล daily_2025 ที่ถูกกรองแล้ว:")
display(daily_2025.head())

จำนวนสถานีก่อนกรอง: 1514
จำนวนสถานีหลังกรอง: 384

ตัวอย่างข้อมูล daily_2025 ที่ถูกกรองแล้ว:


,rainfall_datetime,rainfall_datetime_calc,rainfall_value,tele_station_id,tele_station_oldcode,measure_datetime,station_code,value
43,2025-08-03 00:00:00+07:00,2025-08-03T07:00:00+07:00,0.0,541.0,THNG,2025-08-02,NaN,NaN
44,2025-08-03 00:00:00+07:00,2025-08-03T07:00:00+07:00,0.8,540.0,VLGE31,2025-08-02,NaN,NaN
45,2025-08-03 00:00:00+07:00,2025-08-03T07:00:00+07:00,0.0,538.0,WGCI,2025-08-02,NaN,NaN
47,2025-08-03 00:00:00+07:00,2025-08-03T07:00:00+07:00,0.4,536.0,RIKK,2025-08-02,NaN,NaN
48,2025-08-03 00:00:00+07:00,2025-08-03T07:00:00+07:00,0.0,535.0,DKTI,2025-08-02,NaN,NaN


สถานี 384 จาก 1514 ที่ตรงกับไฟล์ stations in the model_Hill.csv

#ระบุข้อมูลที่หายไป

In [34]:
#ระบุข้อมูลสูญหาย (record ที่หายไป) ของแต่ละสถานี แสดงจำนวนและสัดส่วน
import pandas as pd

# 1. Prepare Data
daily_2025 = daily_2025.copy()
daily_2025['measure_datetime'] = pd.to_datetime(daily_2025['measure_datetime'])

# 2. Calculate Global Date Range constraints
min_date = daily_2025['measure_datetime'].min()
max_date = daily_2025['measure_datetime'].max()
# Calculate total days in the range (Total Possible Records)
total_possible_records = len(pd.date_range(start=min_date, end=max_date, freq='D'))

# 3. Core Logic: Use GroupBy instead of For Loop
# Group by station and count unique dates present for each
# (This replaces the merge inside the loop)
station_stats = daily_2025.groupby('tele_station_oldcode')['measure_datetime'].nunique().reset_index()
station_stats.columns = ['Station', 'Present_Records']

# 4. Calculate Missing Data
station_stats['Missing Records Count'] = total_possible_records - station_stats['Present_Records']
station_stats['Missing Records Proportion'] = station_stats['Missing Records Count'] / total_possible_records

# 5. Final Formatting to match original output exactly
missing_data_df = station_stats[['Station', 'Missing Records Count', 'Missing Records Proportion']]

# Display results
display(missing_data_df.head())
print(f"\nNumber of stations with missing data: {len(missing_data_df[missing_data_df['Missing Records Count'] > 0])}")

,Station,Missing Records Count,Missing Records Proportion
0,ACRU,0,0.000000
1,BBHN,0,0.000000
2,BBUA,15,0.081522
3,BCNG,0,0.000000
4,BDAR,0,0.000000



Number of stations with missing data: 126


#คัดสถานีที่มีข้อมูลสูญหาย >=20% ออก

In [35]:
#คัดสถานีที่มีข้อมูลสูญหาย >=20% ออก แสดงรายชื่อสถานีที่ถูกคัดออกและจำนวนที่เหลือ
# 1. Create a boolean mask for the condition (Reusable)
# สร้างเงื่อนไขแยกไว้ก่อน เพื่อนำไปใช้ทั้งการหาตัวลบ และหาตัวเก็บ
high_missing_mask = missing_data_df['Missing Records Proportion'] >= 0.20

# 2. Get the list of stations to remove using .loc
# ใช้ .loc เพื่อเข้าถึงข้อมูลที่ตรงตามเงื่อนไขทันที
removed_station_codes = missing_data_df.loc[high_missing_mask, 'Station'].tolist()

# 3. Print the list (Using join instead of loop)
# ใช้ .join() แทน for loop เพื่อรวบโค้ดให้สั้นลงและทำงานเร็วขึ้นสำหรับการแสดงผลข้อความ
print("รายชื่อสถานีที่ถูกคัดออกเนื่องจากมีข้อมูลสูญหาย >= 20%:")
if removed_station_codes:
    print("- " + "\n- ".join(removed_station_codes))
else:
    print("ไม่มีสถานีใดถูกคัดออกในเกณฑ์นี้")

print(f"\nจำนวนสถานีที่ถูกคัดออก: {len(removed_station_codes)}")

# 4. Filter daily_2025 using Positive Selection Logic
# นับจำนวนก่อนทำ
before_removal_count = daily_2025['tele_station_oldcode'].nunique()

# แทนที่จะหาตัวลบแล้วใช้เครื่องหมาย ~ (not) เราหา "รายชื่อสถานีที่ผ่านเกณฑ์" (Good Stations) แทน
# แล้วเลือกเฉพาะข้อมูลที่อยู่ในรายชื่อที่ดี
good_stations = missing_data_df.loc[~high_missing_mask, 'Station'] # สังเกตเครื่องหมาย ~ ที่กลับเงื่อนไขตรงนี้แทน
daily_2025 = daily_2025[daily_2025['tele_station_oldcode'].isin(good_stations)]

after_removal_count = daily_2025['tele_station_oldcode'].nunique()

print(f"จำนวนสถานีก่อนคัดออก: {before_removal_count}")
print(f"จำนวนสถานีที่เหลือหลังจากคัดออก: {after_removal_count}")

รายชื่อสถานีที่ถูกคัดออกเนื่องจากมีข้อมูลสูญหาย >= 20%:
- BKHN
- BKK017
- BLKO
- BNGR
- BNLG
- BPTG
- BSKJ
- BSPN
- CEHM
- CGKM
- CHI010
- CHI013
- CHNA
- CPY010
- CTKN
- DCCT
- DGSL
- KKTO
- KON005
- LMSK
- MEWG
- MMMO
- MNRM
- NAPK
- NASW
- NEAK
- NGCG
- NGPI
- NKSG
- NNPT
- NPI002
- PAS008
- PMPN
- PRN001
- PSLT
- SGKW
- SKE001
- SKM004
- SLA002
- TPI007
- TUKW
- VLGE33
- WAN005
- YCUM
- YOM001

จำนวนสถานีที่ถูกคัดออก: 45
จำนวนสถานีก่อนคัดออก: 384
จำนวนสถานีที่เหลือหลังจากคัดออก: 339


#เติมข้อมูลที่หายไปด้วยค่า Median


In [37]:
#เติมข้อมูลที่หายไปด้วยค่า Median ตาม Year จาก historical แยกตามสถานี
# 1. Prepare Historical Median Data
historical['measure_datetime'] = pd.to_datetime(historical['measure_datetime'])

# --- แก้ไขจุดที่ทำให้เกิด Error: สร้างคอลัมน์ doy ให้ชัดเจนก่อน Group ---
historical['doy'] = historical['measure_datetime'].dt.dayofyear
median_ref = historical.groupby(['station_code', 'doy'])['data'].median().reset_index()
# ------------------------------------------------------------------

# Rename columns immediately (เตรียม Merge)
median_ref.rename(columns={'station_code': 'tele_station_oldcode', 'data': 'median_fill_value'}, inplace=True)

# -----------------------------------------------------------------------------
# Optimized Logic: Use 'Reindex' instead of 'Merge with Cartesian Product'
# -----------------------------------------------------------------------------

# 2. Setup the Full Grid (MultiIndex)
unique_stations = daily_2025['tele_station_oldcode'].unique()
date_range = pd.date_range(start=daily_2025['measure_datetime'].min(),
                           end=daily_2025['measure_datetime'].max(),
                           freq='D')

full_idx = pd.MultiIndex.from_product(
    [unique_stations, date_range],
    names=['tele_station_oldcode', 'measure_datetime']
)

# 3. Expand Data using Reindex
daily_2025_imputed = (
    daily_2025
    .drop(columns=['doy'], errors='ignore')
    .set_index(['tele_station_oldcode', 'measure_datetime'])
    .reindex(full_idx)
    .reset_index()
)

# 4. Impute Missing Values
# คำนวณ DOY ใหม่สำหรับข้อมูล daily_2025_imputed
daily_2025_imputed['doy'] = daily_2025_imputed['measure_datetime'].dt.dayofyear

# ตอนนี้ทั้ง daily_2025_imputed และ median_ref มีคอลัมน์ 'doy' ทั้งคู่แล้ว Merge ได้แน่นอน
daily_2025_imputed = pd.merge(
    daily_2025_imputed,
    median_ref,
    on=['tele_station_oldcode', 'doy'],
    how='left'
)

# Fillna: ใช้ค่าจาก median_fill_value ไปอุดรูรั่ว
daily_2025_imputed['rainfall_value'] = daily_2025_imputed['rainfall_value'].fillna(daily_2025_imputed['median_fill_value'])

# 5. Cleanup
daily_2025_imputed.drop(columns=['doy', 'median_fill_value'], inplace=True)

display(daily_2025_imputed.head())

,tele_station_oldcode,measure_datetime,rainfall_datetime,rainfall_datetime_calc,rainfall_value,tele_station_id,station_code,value
0,THNG,2025-07-01,2025-07-02 00:00:00+07:00,2025-07-02T07:00:00+07:00,0.0,541.0,NaN,NaN
1,THNG,2025-07-02,2025-07-03 00:00:00+07:00,2025-07-03T07:00:00+07:00,0.0,541.0,NaN,NaN
2,THNG,2025-07-03,2025-07-04 00:00:00+07:00,2025-07-04T07:00:00+07:00,0.0,541.0,NaN,NaN
3,THNG,2025-07-04,2025-07-05 00:00:00+07:00,2025-07-05T07:00:00+07:00,0.0,541.0,NaN,NaN
4,THNG,2025-07-05,2025-07-06 00:00:00+07:00,2025-07-06T07:00:00+07:00,0.0,541.0,NaN,NaN


## เหตุผลที่ใช้ Median แทน Mean
* **ข้อมูลปริมาณน้ำฝนมักมีการกระจายตัวที่ไม่สมมาตร** (Skewed Distribution): ข้อมูลปริมาณน้ำฝนส่วนใหญ่มักจะมีค่าเป็นศูนย์ (ไม่มีฝนตก) หรือมีค่าต่ำๆ จำนวนมาก และมีค่าสูงๆ (ฝนตกหนัก) เป็นส่วนน้อย ซึ่งทำให้การกระจายตัวของข้อมูลเอียงไปทางขวา (positively skewed).

* **ค่าเฉลี่ย** (Mean) อ่อนไหวต่อค่าผิดปกติ (Outliers) และการกระจายตัวที่ไม่สมมาตร: ถ้ามีค่าฝนตกหนักมากๆ เพียงไม่กี่วัน ค่าเฉลี่ยจะถูกดึงให้สูงขึ้นอย่างมาก ซึ่งอาจไม่สะท้อนถึงปริมาณน้ำฝน 'โดยทั่วไป' ที่สถานีนั้นๆ ได้รับ

* **ค่ามัธยฐาน** (Median) มีความทนทานต่อค่าผิดปกติ (Robust to Outliers): ค่ามัธยฐานคือค่ากลางที่แบ่งข้อมูลออกเป็นสองส่วนเท่าๆ กัน (50% ของข้อมูลมีค่าน้อยกว่าหรือเท่ากับค่ามัธยฐาน และ 50% มีค่ามากกว่าหรือเท่ากับ) การที่ข้อมูลมีค่าสูงหรือต่ำผิดปกติจะไม่ส่งผลกระทบต่อค่ามัธยฐานมากเท่ากับค่าเฉลี่ย ทำให้ค่ามัธยฐานเป็นตัวแทนของ 'ค่าทั่วไป' หรือ 'ค่าปกติ' ของปริมาณน้ำฝนได้ดีกว่าในสถานการณ์เช่นนี้

#คำนวณผลรวม

In [43]:
#คำนวณผลรวมปริมาณน้ำฝนรายเดือน
# แปลงเป็น datetime ให้เรียบร้อย
daily_2025_imputed['measure_datetime'] = pd.to_datetime(daily_2025_imputed['measure_datetime'])

# เริ่ม Pipeline: สร้าง, รวม, เปลี่ยนชื่อ, จัดเรียง ในคำสั่งเดียว
output_monthly_rainfall = (
    daily_2025_imputed
    .assign(
        year=lambda x: x['measure_datetime'].dt.year,
        month=lambda x: x['measure_datetime'].dt.month
    )
    .groupby(['tele_station_oldcode', 'year', 'month'], as_index=False)
    .agg(rain=('rainfall_value', 'sum')) # Named Aggregation: รวมผลแล้วตั้งชื่อว่า 'rain' ทันที
    .rename(columns={'tele_station_oldcode': 'Node'})
    .sort_values(['Node', 'year', 'month'])
    [['Node', 'rain', 'month', 'year']] # จัดลำดับคอลัมน์ปิดท้าย
)

display(output_monthly_rainfall.head())

,Node,rain,month,year
0,ACRU,83.0,7,2025
1,ACRU,164.8,8,2025
2,ACRU,248.6,9,2025
3,ACRU,48.4,10,2025
4,ACRU,52.2,11,2025


In [44]:
output_path = '/content/drive/MyDrive/BSC DPDM midterm 2025 663020277-7/output_monthly_rainfall.csv'
output_monthly_rainfall.to_csv(output_path, index=False)
print(f"บันทึกข้อมูล monthly_rain เป็นไฟล์ CSV ที่: {output_path}")

บันทึกข้อมูล monthly_rain เป็นไฟล์ CSV ที่: /content/drive/MyDrive/BSC DPDM midterm 2025 663020277-7/output_monthly_rainfall.csv



**Prompt AI**: https://docs.google.com/document/d/14jp3P2_bRhTYh8eZsBpHCI_Dx5uLV1w_WlhYeHjInyM/edit?usp=sharing

**Output**: https://drive.google.com/file/d/1VW2QAYkjjXteqFQdwPeR9BT6JRkBnP7-/view?usp=sharing